### ISSUE TITLE
[$250] mWeb - Compose box- Selected text isn't applied format

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#53145]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.67-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**N/A
**Email or phone of affected tester (no customers):** moxtrix+05052023@gmail.com

## Action Performed:
1. Open https://staging.new.expensify.com/
2. At compose box of any conversation, type 1 line
3. Select all text, choose Format: Bold and send message
4. Type another line
5. Select all text, choose Format: Italic and send message
6. Type another line
7. Select all text, choose Format: Underline and send message

## Expected Result:
Selected texts should apply choosing format

## Actual Result:
Selected texts isn't applied choosing format

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [x] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/40872591-76a4-4b44-9383-0249f7fd97a4

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021861616540479439769</li>
        <li>Upwork Job ID: 1861616540479439769</li>
        <li>Last Price Increase: 2024-11-27</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @NicMendonca</details>

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Selected text isn't formatted when selecting format option from the context menu.

### What is the root cause of that problem?
When we have an element with `contenteditable`, the context menu will show the "Format" option. However, the live markdown currently doesn't handle the format command.

### What changes do you think we should make in order to solve the problem?

There are 3 "Format" options, Bold, Italic, and Underline. Currently, we don't have a markdown for Underline. To handle the Bold and Italic format, we need to add new cases for bold and italic format command here.
https://github.com/Expensify/react-native-live-markdown/blob/cec2feae7463398b8d377a231be8671e1698e4a0/src/MarkdownTextInput.web.tsx#L349-L358

```
case 'formatBold':
  newInputUpdate = parseText(
    divRef.current,
    `${parsedText.slice(0, contentSelection.current.start)}*${parsedText.slice(contentSelection.current.start, contentSelection.current.end)}*${parsedText.slice(contentSelection.current.end)}`,
    processedMarkdownStyle,
    newCursorPosition + 2,
    true
  );
  break
case 'formatItalic':
  newInputUpdate = parseText(
    divRef.current,
    `${parsedText.slice(0, contentSelection.current.start)}_${parsedText.slice(contentSelection.current.start, contentSelection.current.end)}_${parsedText.slice(contentSelection.current.end)}`,
    processedMarkdownStyle,
    newCursorPosition + 2,
    true
  );
  break
```

The way the format logic work is by "wrapping" the selected text with the markdown (* for bold, _ for italic). The new selection will be added by 2 because we add 2 * or _. 

But the Underline format still won't do anything because we don't have a markdown for that, and because the live markdown also handles the styling manually, so the alternative is to disable the "Format" option. We can do that by setting the `contenteditable` to "plaintext-only"
https://github.com/Expensify/react-native-live-markdown/blob/cec2feae7463398b8d377a231be8671e1698e4a0/src/MarkdownTextInput.web.tsx#L724

```
contentEditable: disabled ? 'false' : 'plaintext-only',
```

Based on the doc, "plaintext-only" is available on most browsers except Firefox, but this Format context menu is only available on Safari, so I think it's not a problem.
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/contenteditable#browser_compatibility

**Comment by kadiealexander** [comment]
Bumped @mananjadhav for review [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1733176854688559).

**Comment by mananjadhav** [comment]
I checked @bernhardoj's proposal. I think we can go ahead with this.  But before we finalize I want to understand if we could remove the Underline option from the Format menu? (Atleast on Whatsapp I don't see it), whereas Slack removes the format option for me.

Nevertheless we can use @bernhardoj's proposal as they've mentioned the alternative for removing the `Format` menu as well.

Tagging @Expensify/design for the input as well.

🎀 👀 🎀 C+ reviewed.

**Comment by shawnborton** [comment]
Oh interesting, I didn't even know you could do this from iOS. I suppose we could make it so that whatever style you picked from the native menu (bold, italic, etc) was applied via our markdown styles, as in we'd prepend and append a * if you picked bold. Is something like that possible?

**Comment by dubielzyk-expensify** [comment]
That'd be great if that's possible. If not, I wonder if we could hide or turn off that formatting menu altogether.

**Comment by dannymcclain** [comment]
I'm fine with making just bold and italic work or hiding that format menu item altogether. I also didn't even realize it was there haha.

**Comment by bernhardoj** [comment]
Thanks for the confirmation. Opened the PR on live markdown repo here: https://github.com/Expensify/react-native-live-markdown/pull/556

Waiting for the review from the live markdown team.

**Comment by mananjadhav** [comment]
> However, this PR only implements Format menu on web. I believe the same feature is missing on native iOS. Are there any plans to implement this missing feature also in native iOS?

@bernhardoj That's a good question raised [here](https://github.com/Expensify/react-native-live-markdown/pull/556#issuecomment-2517909791). Do we not need it for native?

**Comment by bernhardoj** [comment]
Replied on the PR

**Comment by mananjadhav** [comment]
Keeping the conversation here to understand the scope.  @bernhardoj I can see `Format` option in the native iOS Whatsapp. When clicked these are the options I see. This is what I also meant if we could hide just the underline. This translates to markdown btw.

![format-native-whatsapp](https://github.com/user-attachments/assets/329c1777-7168-400f-b189-8e315c57ccaf)

**Comment by bernhardoj** [comment]
But it's not available on our app, so I'm not sure how we can handle it.

**Comment by blimpich** [comment]
Issue is assigned, automation didn't work quite right for some reason. Conversation is active in the PR.

**Comment by mananjadhav** [comment]
The conversation is the in Expensify/react-native-live-markdown repo. May be that's why.

**Comment by blimpich** [comment]
Looks like we're waiting on reviews on the PR @bernhardoj @mananjadhav?

**Comment by bernhardoj** [comment]
Yes, I bumped on the PR

**Comment by dubielzyk-expensify** [comment]
Not overdue

**Comment by blimpich** [comment]
Waiting on reviews, bumped the agency reviewers in the PR.

**Comment by blimpich** [comment]
Still waiting on reviews. Holidays are slowing down the review process for this one I think.

**Comment by blimpich** [comment]
Fantastic, looks like react-native-live-markdown has been updated and the npm package published. @bernhardoj @mananjadhav we need to update the package in App now right? Then we can call this issue done?

**Comment by bernhardoj** [comment]
App PR is ready

cc: @bernhardoj

**Comment by blimpich** [comment]
PR is merged. Created follow up issue here to add updated behavior: https://github.com/Expensify/App/issues/55088.

**Comment by mananjadhav** [comment]
@blimpich Is it possible to assign the issue to me for the review?

**Comment by blimpich** [comment]
@mananjadhav sure! I'll assign you to review once we're ready to add the `external` label

**Comment by mananjadhav** [comment]
Okay thanks @blimpich

**Comment by blimpich** [comment]
@bernhardoj this follow up issue is open to proposals now in case you're interested: https://github.com/Expensify/App/issues/55088

**Comment by bernhardoj** [comment]
@blimpich another reason it gets reverted is because there is an [issue](https://github.com/Expensify/App/issues/55115) with the live markdown on a version between 0.1.210 
(current App version) and 0.1.215 (our PR version). Looks like it's related to https://github.com/Expensify/App/issues/14676 (most likely the 0.1.214 version) and currently being handled in this live-markdown [PR](https://github.com/Expensify/react-native-live-markdown/pull/597).

Until it's fixed, we can't update it yet. (and it's better for them to handle the live-markdown update)

**Comment by blimpich** [comment]
I see, thank you for the added context @bernhardoj! Lets wait on https://github.com/Expensify/App/issues/55115 to be finished then before moving forward. I'll add a hold and mark this as weekly.

**Comment by kadiealexander** [comment]
@mananjadhav could you please complete the checklist?

**Comment by blimpich** [comment]
@kadiealexander the PR ended up being reverted and this issue is now on hold, waiting for another issue to be completed first, so we need to wait to re-merge the PR before going through the bugzero checklist.

**Comment by kadiealexander** [comment]
Still waiting on #55115

**Comment by kavimuru** [comment]
underline is not fixed yet. Bold and Itallic are fixed

https://github.com/user-attachments/assets/a9b1e010-4d71-4d9a-ba59-52409d363999

**Comment by mananjadhav** [comment]
I think we can track underline as a follow up. This was discussed earlier that for now we have only Bold and Italic support.

**Comment by blimpich** [comment]
Sounds good. @mananjadhav we still need to pay out this issue correct?

**Comment by mananjadhav** [comment]
Yes, need a payout summary, @kadiealexander. While I'll work on the checklist.

**Comment by mananjadhav** [comment]
# BugZero Checklist:

- [x] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [x] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [x] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [x] 3z. Other:

</details>

- [x] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment: NA

- [x] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion: NA

- [x] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


 Only for iOS mWeb Safari and Web Safari
 
 1. Open any chat
 2. Type anything on the composer
 3. Select some text on the composer
 4. Right-click or long-press to open the web context menu
 5. Select Format (iOS mWeb) or Font (Web Safari)
 6. Select Bold and Italic
 7. Verify the bold and italic markdown is added.

Do we agree 👍 or 👎


</details>

**Comment by mananjadhav** [comment]
@kadiealexander This is all set to be paid out.

**Comment by NicMendonca** [comment]
### Payment Summary

- Contributor: @bernhardoj paid $250 via Upwork
- Contributor+: @mananjadhav due $250 via NewDot

**Comment by NicMendonca** [comment]
@bernhardoj I had to re-issue you an offer - can you please accept it when you can? Thanks!

**Comment by bernhardoj** [comment]
@NicMendonca I'll request it through ND.

**Comment by bernhardoj** [comment]
Requested in ND.

**Comment by mananjadhav** [comment]
Thank you @NicMendonca for taking care of this.

**Comment by garrettmknight** [comment]
$250 approved for @mananjadhav

**Comment by NicMendonca** [comment]
### Updated payment summary: 

**Payment Summary**
Contributor: @bernhardoj paid $250 via NewDot
Contributor+: @mananjadhav due $250 via NewDot

**Comment by mananjadhav** [comment]
@NicMendonca We're good to close this one out. Just one regression test proposal [here](https://github.com/Expensify/App/issues/53145#issuecomment-2640224582)

**Comment by JmillsExpensify** [comment]
$250 approved for @bernhardoj
